### PR TITLE
Milestone 3

### PR DESCRIPTION
**Pull request by prolajumokeoni**
## Milestone 3
- Added a friendship model with associations 
- Users can see other posts from his friend
- Users can send, reject or accept a friend request
- Added units and integration tests

### CODE REVIEW

**Review by mke2111 — CHANGES_REQUESTED** [review]
# STATUS: CHANGES REQUIRED 🔴 ♻️ 

## Hey Happy Coder, 👋🏽 
**This is a great project you have going on, well-done**. 👍🏽 

### Well-done 💯 
    1. The tests run just fine.

**However, there are a few sections to improve before your project is approved.** 🔴 
<hr>

### REQUIRED
> You seem to have missed the requirements for this milestone since you haven't made any of the requested changes. I'll explain to you what exactly the requirements need. I won't use comments since the files that need necessary changes aren't available for use.

- [ ] In the User model, once a friendship is confirmed, an additional row with reversed user-friend pair is created using the `confirm_friend(user)` method. Thanks to this, we can find user friends only by user_id in friendships table and can be defined as association with SQL query instead of processing objects in application memory (as the database was created to store and query data it is always more efficient than ruby)
```
# User model

has_many :confirmed_friendships, -> { where confirmed: true }, class_name: "Friendship"
has_many :friends, through: :confirmed_friendships
```

- [ ] Using the same approach as above and for the same reason, change the `pending_friends` method. `pending_friends` can be defined as association with SQL query instead of processing objects in application memory (as the database was created to store and query data it is always more efficient than ruby)

> You are currently using a ruby enumerable placed in the helper catering for both the pending and the confirm features

- [ ] Now you can guess what's next, 😄 `friend_requests` method.
`friend_requests` can be defined as association with SQL query instead of processing objects in application memory (as the database was created to store and query data it is always more efficient than ruby). Remember to use 'has_many' & 'through' ':inverted_friendships'

- [ ] Users Controller index action should display all users with Friend Request button next to users who are not their friends already

### OPTIONAL

- [ ] It is a good practice to have zero logic in the views directory. Create methods in the helpers directory and call the methods in the views directory instead.
Read this [article](https://medium.com/@acushla4real/rails-view-helpers-best-practices-guide-d7d7b6db3db4) for a head-start. Another great piece to read is this [article](https://brandonhilkert.com/blog/refactoring-logic-from-a-rails-view/)

#

### Kindly make all the requested changes and submit another CRR. 💯 

**Cheers 🍹 and Happy coding 💻** 

<hr>

Mention @mke2111 in this thread and I'll be glad to help. 😄 
### Best regards
Slack 🆘  Roy Mukuye
Twitter 🐦  [Roy Mukuye](https://twitter.com/Roymkenya)
LinkedIn 🔗  [Roy Mukuye](https://www.linkedin.com/in/roy-mukuye/)

**Review by mke2111 — CHANGES_REQUESTED** [review]
# STATUS: CHANGES REQUIRED 🔴 ♻️ 

## Hey Happy Coder, 👋🏽 
**Thanks for the efforts of trying to implement the changes requested earlier on**. 

### Well-done 💯 
    1. The app runs almost as expected.

**However, there are a few sections to improve before your project is approved.** 🔴 
<hr>

### REQUIRED

- [ ] Make all the changes previously requested. If you need more help, we can schedule another meeting. 
#

### Kindly make all the requested changes and submit another CRR. 💯 

**Cheers 🍹 and Happy coding 💻** 

<hr>

Mention @mke2111 in this thread and I'll be glad to help. 😄 
### Best regards
Slack 🆘  Roy Mukuye
Twitter 🐦  [Roy Mukuye](https://twitter.com/Roymkenya)
LinkedIn 🔗  [Roy Mukuye](https://www.linkedin.com/in/roy-mukuye/)

**Comment by prolajumokeoni** [comment]
> # STATUS: CHANGES REQUIRED  
> ## Hey Happy Coder, 👋🏽
> **Thanks for the efforts of trying to implement the changes requested earlier on**.
> 
> ### Well-done 
> ```
> 1. The app runs almost as expected.
> ```
> 
> **However, there are a few sections to improve before your project is approved.** 
> 
> ### REQUIRED
> * [ ]  Make all the changes previously requested. If you need more help, we can schedule another meeting.
> 
> ### Kindly make all the requested changes and submit another CRR. 
> **Cheers tropical_drink and Happy coding computer**
> 
> Mention @mke2111 in this thread and I'll be glad to help. 
> 
> ### Best regards
> Slack  Roy Mukuye
> Twitter  [Roy Mukuye](https://twitter.com/Roymkenya)
> LinkedIn  [Roy Mukuye](https://www.linkedin.com/in/roy-mukuye/)

I would appreciate another meeting, when are you available? @mke2111

**Comment by prolajumokeoni** [comment]
I tried implementing the changes but I am getting a lot of errors, is it mandatory to do it this way, because it will require me to restart the project afresh which will interfere with milestone 2 which is already approved.

File: app/controllers/users_controller.rb [hunk 1]
```diff
@@ -46,3 +46,14 @@ def destroy
...
+def friends?(user)
+  Friendship.exists?(user_id: current_user.id, friend_id: user)
+end
+
+def sent_requests
+  @requested_friends = []
+  @sent_requests = Friendship.where('user_id = ? and status = ?', current_user.id, false)
+  @sent_requests.each { |u| @requested_friends << User.find(u.friend_id) }
+  @requested_friends
+end
```
Comment (carlos-ssh) [inline app/controllers/users_controller.rb:59]: **Kindly fix this:**
- When i have a friendship i can send  a friend request to a person who  added me before.
- Kindly add a button to cancel a friend request.
- I can't remove a friend request.
- Into your all users page add a button for friend request.

[OPTIONAL]

- Add your Author and a license into your README file.

**Review by carlos-ssh — CHANGES_REQUESTED** [review]
Hi @prolajumokeoni , 

Great work on making the changes requested by a previous reviewer 👏 
You've done well implementing some of the requested changes, but there are still some which aren't addressed yet.


 ## Required Changes ♻️
_Check the comments under the review._

## Optional suggestions

_Every comment with the [OPTIONAL] prefix won't stop the approval of this PR. However, I strongly recommend you to take them into account as they can make your code better. Some of them were simply missed by the previous reviewer and addressing them will really improve your application._


Cheers and Happy coding!👏👏👏

Feel free to leave any questions or comments in the PR thread if something is not 100% clear.

**_Please, do not open a new Pull Request for re-reviews. You should use the same Pull Request submitted for the first review, either valid or invalid unless it is requested otherwise._**


------
_As described in the [Code reviews limits policy](https://microverse.zendesk.com/hc/en-us/articles/1500004088561) you have only 3 code reviews per project. If you think that the code review was not fair, you can request a second opinion using [this form](https://airtable.com/shrQAqnBwek5a0O0s)._

File: app/controllers/users_controller.rb (hunk 1 above)
Comment (prolajumokeoni) [reply app/controllers/users_controller.rb:59]: @carlos-ssh all changes have been implemented.

**Review by george-shammar — APPROVED** [review]
### **STATUS: APPROVED** :heavy_check_mark: :heavy_check_mark: 

**Impressive**!!!! :muscle: :clap: 

**Highlight**
:heavy_check_mark: You've implemented the required changes from the last review. 

You may go ahead to merge this branch and proceed to the next milestone.

I wish you good luck.

**Cheers** :smile: **With respect** :tophat:

**Pull request merged**